clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -922,6 +946,103 @@ public void testTransactionManagerEnablesV2() {
         assertTrue(transactionManager.isTransactionV2Enabled());
     }
 
+    @Test
+    public void testTransactionV2AddPartitionAndOffsets() throws InterruptedException {
+        initializeTransactionManager(Optional.of(transactionalId), true);
+        doInitTransactions();
+
+        transactionManager.beginTransaction();
+
+        Future<RecordMetadata> responseFuture = appendToAccumulator(tp0);
+
+        assertFalse(responseFuture.isDone());
+
+        prepareProduceResponse(Errors.NONE, producerId, epoch);
+        assertTrue(transactionManager.transactionContainsPartition(tp0));
+        assertTrue(transactionManager.isSendToPartitionAllowed(tp0));
+        assertFalse(responseFuture.isDone());
+        runUntil(responseFuture::isDone);
+
+        // Now, test adding the offsets.
+        Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
+        offsets.put(tp1, new OffsetAndMetadata(1));
+
+        TransactionalRequestResult addOffsetsResult = transactionManager.sendOffsetsToTransaction(
+            offsets, new ConsumerGroupMetadata(consumerGroupId));
+
+        assertTrue(transactionManager.hasPendingOffsetCommits());
+
+        // the result doesn't complete until TxnOffsetCommit returns
```

Comment:
Do we consider the offset partitions added to the transaction before the response returns?

Reply:
Looks like we don't have a "offset partition added" state in the old code. We always call `AddOffsetsToTxnRequest` before sending offsets.

Reply:
Interesting. I wonder if this could cause some weird behavior in the case where we only try to send offset commits. We had a lot of discussion about what happens if we don't know if we added a partition, but I guess here it is ok since we have to go through two operations through the txn, 1 to add and 2 to commit an offset.

Reply:
Ending up discussing this a bit with Artem offline a little while back. This could cause an issue if we didn't allow empty -> prepareAbort transition, but since we have that (and its variants) this won't be a problem here either.  

Essentially we always waited for addOffsetsToTxn to return before sending the offset, so we wouldn't be in a limbo state about the transaction starting (we set transactionStarted true when the offset is added) but now we don't have that, so we require the ability to abort on the other states in the same way.